sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala

```diff
@@ -276,6 +276,97 @@ abstract class HashExpression[E] extends Expression {
     }
   }
 
+  protected def genHashInt(i: String, hasher: String, result: String): String =
```

Comment:
The hasher should be a static class. Show we don't really need to pass it around. We could use  `hasherClassName` instead.